overlord/state/taskrunner.go

```diff
@@ -78,16 +78,18 @@ func taskFail(t *Task) {
 // run must be called with the state lock in place
 func (r *TaskRunner) run(fn HandlerFunc, task *Task) {
 	task.SetStatus(RunningStatus) // could have been set to waiting
-	tomb := &tomb.Tomb{}
-	r.tombs[task.ID()] = tomb
-	tomb.Go(func() error {
-		err := fn(task, tomb)
+	tb := &tomb.Tomb{}
+	r.tombs[task.ID()] = tb
+	tb.Go(func() error {
+		err := fn(task, tb)
 
 		r.state.Lock()
 		defer r.state.Unlock()
 		halted := task.HaltTasks()
-		if err == nil {
+		if err == nil && tb.Err() == tomb.ErrStillAlive {
```

Comment:
Sorry, this comment is in the wrong location in the file changes pane. Please see the conversation pane for the right context (bug reported to GitHub).